### ISSUE TITLE
SC2154 invalid global variable match

### ISSUE THREAD

**Issue by bruce-ricard** [koalaman/shellcheck#2904]
#### For bugs
- Rule Id : SC2154
- My shellcheck version: `0.9.0` 
- [x] The rule's wiki page does not already cover this (e.g. https://shellcheck.net/wiki/SC2086)
- [ ] I tried on https://www.shellcheck.net/: the web UI doesn't appear to run this check

#### Here's a snippet or screenshot that shows the problem:

```sh

#!/bin/bash

set -u
echo "${GLOBAL_VARIABLE}"

```

#### Here's what shellcheck currently says:

When checking it with `shellcheck --enable=all`, we get:

> SC2154 (warning): GLOBAL_VARIABLE is referenced but not assigned.

#### Here's what I wanted or expected to see:

No warning, as the [documentation](https://github.com/koalaman/shellcheck/wiki/SC2154) states
>  This message only triggers for variables with lowercase characters in their name (foo and kFOO but not FOO) due to the standard convention of using lowercase variable names for unexported, local variables.

**Comment by bruce-ricard** [comment]
I've got new data. It appears that the issue is not with SC2154, but with  the `check-unassigned-uppercase` check, for which I don't know the code. Because when you run `shellcheck  --enable=check-unassigned-uppercase`, it displays the error message for SC2154, which appears not to be the correct one.

**Comment by brother** [comment]
For your example code above I would consider SC2154 appropriate. You try to use a variable that is not assigned in a context known to shellcheck.

**Comment by bruce-ricard** [comment]
> For your example code above I would consider SC2154 appropriate. You try to use a variable that is not assigned in a context known to shellcheck.

The documentation states that it doesn't check all caps global variables. Something's wrong.

> a variable that is not assigned in a context known to shellcheck

We cannot start checking for this. Otherwise we also need to warn about

```sh
ls
```

with: WARNING, the function `ls` may or may not be defined in this context.

**Comment by bruce-ricard** [comment]
This check is equivalent to warn about

```
function add(int x, int y) {
   return x + y;
}
```

with "WARNING, `x` and `y` may be undefined. They are not set anywhere in this file!"

**Comment by brother** [comment]
Chill. It's hard to follow when you change the OP.

```shell

brother /tmp/test$ cat example.sh 
#!/bin/bash


echo "${GLOBAL_VARIABLE}"
brother /tmp/test$ shellcheck example.sh 
brother /tmp/test$ echo $?
0
brother /tmp/test$ shellcheck --enable=check-unassigned-uppercase example.sh 

In example.sh line 4:
echo "${GLOBAL_VARIABLE}"
      ^----------------^ SC2154 (warning): GLOBAL_VARIABLE is referenced but not assigned.

For more information:
  https://www.shellcheck.net/wiki/SC2154 -- GLOBAL_VARIABLE is referenced but...
brother /tmp/test$ echo $?
1
```

SC2154 will warn about usage of uninitialized variables. The **default** behavior is to **skip upper case** because they are considered to be global from outside the script.
If you explicitly enable the check to include the upper case variables by `--enable=check-unassigned-uppercase` or via `--enable=all`. it will ofc trigger for upper case as well.

There are examples of developers not knowing the scope handling of bash and they put something that looks like instance variables in upper case at the top to define certain things and then use them as "global for the script". In that style of code you need to include them in the check, hence `--enable=check-unassigned-uppercase`.
Enable the check and vet the use, remove errors found. And then leave it default disabled.

**Comment by brother** [comment]
`--enable=` is used to enable a list of optional checks. There are currently nine extra checks not enabled by default.
If you enable a optional check it can alter the default behavior, this is true for SC2154 with `check-unassigned-uppercase`. 

SC2154 is a default check. You can disable a default check with `--exclude=SC2154`.

You can also provide a list of codes to check with `--include=SC2154`,  that would disable all the other checks.

**Comment by brother** [comment]
> When I ran `shellcheck --enable=all`, I got this warning message, telling me that uppercase variables are not included in this check, despite my variable being all uppercase. I had no idea that there existed a check that also validates global variables. That is a bug.
> I don't know the relationship between `check-unassigned-uppercase` and `SC2154`. AFAICT they are not the same check. One is about local variables, the other about global ones.

I t works as intended.

By default, without any `--enable` at all, the uppercase will not get flagged.
If you provide either `all` or `check-unassigned-uppercase` it will be flagged.

It is hard to cater for everyone using the tool but most people finding the `--enable=all` option is also capable of finding the list of the checks.
This behavior documented both in `--list-optional` and the wiki. If you want to change the wording of the wiki you can and are welcome to.

**Comment by bruce-ricard** [comment]
I agree that the feature works as intended. I understand that the code is flagged when you pass these parameters.

What I'm saying is that the error message and the documentation is not the one associated with that error.

I now understand that I am seeing the appropriate warning. But at the beginning, because the message was so confusing, I thought that the warning was a bug. Now I understand that the problem is the documentation. 
What I'm saying is that people will keep getting confused by the warning if we don't fix the documentation.

>If you want to change the wording of the wiki you can and are welcome to.

I don't think that this will do. It seems to me like the issue is that the same SC warning is issued for both uppercase and lowercase variables. 

I guess we could try to add to the wiki page something like: "by the way, if you happened to get this warning for your all uppercase external variables, it's because you ran with `--enable=all` or your enabled the uppercase check. In this case, the note doesn't apply to you, and the uppercase variables are correctly flagged by ShellCheck".
It doesn't look great to me, but it would at least share the information.
